src/main/scala/io/iohk/ethereum/ledger/BlockPreparator.scala

```diff
@@ -0,0 +1,307 @@
+package io.iohk.ethereum.ledger
+
+import io.iohk.ethereum.consensus.validators.SignedTransactionValidator
+import io.iohk.ethereum.domain.UInt256._
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.BlockExecutionError.{StateBeforeFailure, TxsExecutionError}
+import io.iohk.ethereum.ledger.Ledger._
+import io.iohk.ethereum.utils.{BlockchainConfig, Logger}
+import io.iohk.ethereum.vm.{PC ⇒ _, _}
+
+import scala.annotation.tailrec
+
+/**
+ * This is used from a [[io.iohk.ethereum.consensus.blocks.BlockGenerator BlockGenerator]].
+ *
+ * With its introduction, we:
+ *
+ *   1. Avoid a direct dependency of [[io.iohk.ethereum.consensus.Consensus Consensus]] on
+ *      [[io.iohk.ethereum.ledger.Ledger Ledger]].
+ *   2. Extract a substantial chunk of functionality outside [[io.iohk.ethereum.ledger.Ledger Ledger]],
+ *      in an attempt to modularize it.
+ *
+ */
+class BlockPreparator(
```

Comment:
I was wondering what's the difference between block preparation and block generation. I can see `BlockPreparator` handles block execution so why not `BlockExecutor`? 

But from the doc I understood that it was meant to take that responsibility from `Ledger`. In the `Ledger` however, methods like `executeTransactions` still exist, so I'm a bit confused. Why the duplication?

Reply:
I named it `BlockPreparator` because of its `prepareBlock` method. This is a dependency that `Consensus` would have on `Ledger`, and the one I wanted to avoid. So I started by taking this out of `Ledger` and so I created  a class named `BlockPreparator` because it had this important `prepareBlock` method. The remaining functionality that you see in `BlockPreparator` is the transitive dependencies needed in order to fully implement `prepareBlock`. These were taken from `Ledger` as well. But I did not remove the respective methods from `Ledger`, I only delegated to `BlockPreparator`. The reason for that is because these methods were already used in tests and I did not want to be carried away to a more massive refactoring.

HTH

Reply:
I see, maybe we should a task to refactor/clean up those tests. For example the method `LedgerImpl#executePreparedTransactions` is only used in tests, so the avoidance of refactoring the tests is it's only reason to exist.

As for the name, it seems that `BlockPreparator` is currently used for 2 things: block preparation and block execution (of already prepared blocks, like those received from peers). Because preparation includes and depends on execution, I would say it's more appropriate to name it `BlockExecutor` or `TxExecutor`. It's more OK if `*Executor` has an extra functionality that is preparation, rather `Preparator` be used mainly for something other than preparation. Note that `LedgerImp#prepareBlock` is also only used in tests, so it probably  should be removed, and the tests refactored.

Reply:
I mean, no code changes required here, but tell me what do you think about refactoring the tests as a follow-up. Is it worth the effort?

Reply:
I have no doubt we must attempt to refactor the tests in a a post-PR follow-up action.